### ISSUE TITLE
Reduce false positives caused by sunset hues and dark clouds

### ISSUE THREAD

**Issue by dhyan2815** [dhyan2815/SmokeSignal-AI#12]
> ### Task 5: User Testing/Feedback Documentation Complete ✅
> 
> I have completed Task 5 of the implementation roadmap by conducting a beta test and documenting the user feedback.
> 
> **Artifacts Created:**
> - \docs/TESTING.md\: Comprehensive report analyzing feedback from 6 beta testers via Google Forms.
> 
> **Key Results & Metrics:**
> - **UI/UX Rating**: 5.0/5.0 average for responsiveness and ease of use.
> - **Overall Experience**: 4.83/5.0 average rating.
> - **Accuracy**: 100% accuracy on standard wildfire/smoke images.
> 
> ### ⚠️ Model Challenges Identified
> During beta testing, users discovered specific edge cases that triggered false alarms:
> - **Sunset/Sunrise Hues**: The model misinterprets red/orange sunset skies as active wildfires.
> - **Dark/Monsoon Clouds**: Thick, dark clouds can also trigger false positives.
> 
> ### 💡 Future Enhancements Roadmap
> Based on user feedback, the following features have been logged for future implementation:
> - **Model Enhancements**: Retrain with sunset/sunrise images to reduce false alarms; transition to a multi-class model (smoke, clouds, fog).
> - **Platform Features**: Map/location integration, automated Email/SMS alerts, a logging database for detections, and a one-click emergency contact button. 

 _Originally posted by @dhyan2815 in [#10](https://github.com/dhyan2815/SmokeSignal-AI/issues/10#issuecomment-4639803787)_

## Current behavior / bottleneck

During beta testing, the wildfire detection model produced false positive alerts when processing scenes containing sunset/sunrise sky hues and thick dark or monsoon-type clouds. The CNN misclassifies red/orange atmospheric coloration and low-luminance cloud formations as active wildfire events. The current binary classification architecture (fire / no-fire) lacks the granularity to distinguish between visually similar phenomena, making the model unreliable under these common real-world conditions.

## Expected improvement

False positive rate for sunset/sunrise and dark cloud scenes should drop to below 5% on a held-out validation set representative of these edge cases. Overall precision on the existing benchmark dataset should be maintained or improved. The model should correctly classify smoke, clouds, and fog as distinct non-fire categories rather than collapsing them into a single negative class.

## Proposed approach

- Curate and label an augmented training dataset that includes sunset/sunrise sky images, monsoon cloud formations, fog, and wildfire smoke as explicit classes.
- Transition the classification head from binary output to a multi-class softmax output (e.g., `fire`, `smoke`, `cloud`, `fog`, `clear`).
- Retrain the CNN backbone (with optional transfer learning from the current checkpoint) on the expanded dataset, applying class-balanced sampling or weighted cross-entropy loss to handle potential class imbalance.
- Introduce image preprocessing normalization tuned to reduce sensitivity to color temperature shifts (e.g., HSV-space augmentation during training to simulate varying atmospheric lighting).
- Evaluate model performance against a dedicated edge-case test split containing verified sunset, cloud, and fog scenes before any production deployment.
- Update the inference pipeline to surface the predicted class label and confidence score in detection logs and alert payloads, enabling downstream filtering logic.

## Acceptance criteria

- [ ] A labeled edge-case validation set containing at minimum 500 sunset/sunrise images and 500 dark/monsoon cloud images is created and version-controlled in the dataset repository.
- [ ] The retrained model achieves a false positive rate below 5% on the edge-case validation set for both sunset and dark cloud scene categories.
- [ ] The model outputs multi-class predictions (`fire`, `smoke`, `cloud`, `fog`, `clear`) with an associated confidence score.
- [ ] Overall precision and recall on the existing wildfire benchmark dataset does not regress by more than 2% compared to the current production model.
- [ ] The inference pipeline correctly propagates the new class labels and confidence scores to detection logs and alert payloads.
- [ ] Retrained model weights and updated class label mappings are committed and documented in the model registry.

## Related issues

Closes #10

**Priority:** P1 — high

**Comment by dhyan2815** [comment]
# Plan A: Reduce False Positives (Issue #12)

**Goal:** Transition SmokeSignal AI from binary classification (`fire` / `no-fire`) to a 5-class model (`clear`, `cloud`, `fire`, `fog`, `smoke`) to eliminate false positives caused by sunset hues and dark clouds, while maintaining or improving accuracy on the existing wildfire benchmark.

---

## Resolved Decisions

| Decision | Resolution |
|---|---|
| Alert trigger classes | Both `fire` **and** `smoke` trigger email alerts |
| Class mapping index | `0: clear`, `1: cloud`, `2: fire`, `3: fog`, `4: smoke` |
| Training environment | Provide both `src/train_multiclass.py` (local) and `notebooks/SmokeSignal-AI-Multiclass.ipynb` (Colab GPU) |

> [!WARNING]
> **Non-Regression Guarantee**
> The retraining pipeline will include automated regression checks: precision and recall on the existing wildfire benchmark test split must not degrade by more than 2%. Training will use transfer learning from the current checkpoint, early stopping, and model checkpointing to preserve the quality of the existing model while extending its capabilities.

---

## Phase 1 — Dataset Curation & Augmentation

Curate a labeled edge-case dataset meeting the acceptance criteria (≥500 sunset/sunrise images, ≥500 dark/monsoon cloud images) and reorganize the existing dataset into the 5-class folder structure.

#### [NEW] [curate_dataset.py](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/src/curate_dataset.py)
- Downloads and organizes edge-case images from open-source repositories (OpenImages, Hugging Face datasets, or Flickr Public API).
- Creates the folder structure:
  ```
  data/multiclass/
  ├── train/
  │   ├── clear/       # existing "nowildfire" images + new clear-sky/sunset images
  │   ├── cloud/       # new dark/monsoon cloud images
  │   ├── fire/        # existing "wildfire" images
  │   ├── fog/         # new fog/mist images
  │   └── smoke/       # new smoke-only images (no active flame)
  ├── valid/
  │   └── (same subfolders)
  └── test/
      └── (same subfolders)
  ```
- Splits edge-case images into train/valid/test (70/15/15).
- Prints verification summary confirming ≥500 sunset and ≥500 cloud images in the validation set.

#### [NEW] [data/edge_cases/](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/data/edge_cases/)
- Version-controlled directory for the curated edge-case validation images, organized by category (`sunset/`, `cloud/`, `fog/`, `smoke/`).

---

## Phase 2 — Model Architecture & Retraining

Transition the CNN from binary sigmoid output to 5-class softmax, retrain with quality safeguards, and save the updated model.

#### [NEW] [train_multiclass.py](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/src/train_multiclass.py)
- **Transfer learning:** Loads the current checkpoint (`model/wildfire_detector_model.keras`), freezes the convolutional backbone (`Conv2D` + `MaxPooling2D` layers), and replaces the classification head:
  ```
  Dense(1, sigmoid)  →  Dense(5, softmax)
  ```
- **Two-stage training strategy** (to preserve quality):
  1. **Stage 1 — Head-only training:** Train only the new classification head (Dense layers) for ~5 epochs with the backbone frozen. This lets the head learn the 5-class boundaries without disturbing the learned feature extractor.
  2. **Stage 2 — Fine-tuning:** Unfreeze the full model and train end-to-end at a reduced learning rate (`1e-5`) for ~10 epochs with early stopping (patience=3, monitoring `val_loss`).
- **Loss:** `categorical_crossentropy` (one-hot labels via `class_mode='categorical'` in the data generator).
- **Class balancing:** Computes `class_weight` from training set distribution to handle imbalance between the original 30k wildfire/no-wildfire images and the smaller new categories.
- **Data augmentation pipeline** (applied during training):
  - Standard: rotation (±20°), horizontal flip, width/height shift (±10%), zoom (±15%).
  - **HSV color jittering** via a custom `preprocessing_function`: random hue shift (±15°), saturation scale (0.7–1.3), brightness scale (0.7–1.3). This desensitizes the model to sunset-like color temperature shifts.
- **Model checkpointing:** Saves the best model (lowest `val_loss`) to `model/wildfire_detector_multiclass.keras`.
- **Regression check:** After training, automatically evaluates precision/recall on the original binary test split and compares against baseline metrics (96.09% precision, 96.81% recall). Prints a PASS/FAIL verdict.
- Saves the final class label mapping to `model/class_labels.json`.

#### [NEW] [SmokeSignal-AI-Multiclass.ipynb](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/notebooks/SmokeSignal-AI-Multiclass.ipynb)
- Google Colab-optimized notebook mirroring `train_multiclass.py` with:
  - GPU runtime auto-detection and memory growth configuration.
  - Kaggle dataset download cell for the original wildfire dataset.
  - Edge-case dataset upload/download cells.
  - Full training history visualization (accuracy/loss curves, confusion matrix, per-class precision/recall/F1).
  - Model export cell to download the `.keras` file and `class_labels.json`.

---

## Phase 3 — Inference, UI & Alerting Pipeline Updates

Update all downstream components to consume the 5-class model output.

#### [MODIFY] [config.py](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/config.py)
- Add class configuration:
  ```python
  CLASSES = ['clear', 'cloud', 'fire', 'fog', 'smoke']
  ALERT_CLASSES = ['fire', 'smoke']  # classes that trigger email alerts
  MODEL_PATH = "model/wildfire_detector_multiclass.keras"
  ```

---

#### [MODIFY] [app.py](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/app.py)
- **`predict()` function:** Replace the binary sigmoid logic:
  ```python
  # Before (binary)
  confidence_score = float(preds[0][0])
  result = confidence_score > 0.5

  # After (multi-class)
  class_idx = int(np.argmax(preds[0]))
  class_name = Config.CLASSES[class_idx]
  confidence = float(preds[0][class_idx])
  is_alert = class_name in Config.ALERT_CLASSES
  ```
- **UI display:** Show class-specific labels and icons:
  | Predicted Class | UI Label |
  |---|---|
  | `fire` | 🔥 **Wildfire Detected** |
  | `smoke` | 💨 **Smoke Detected** |
  | `cloud` | ☁️ **Clouds Detected — No Wildfire** |
  | `fog` | 🌫️ **Fog Detected — No Wildfire** |
  | `clear` | ✅ **Clear Landscape — No Wildfire** |
- **Confidence bar:** Display a horizontal bar chart showing the softmax probabilities for all 5 classes so the user can see the full distribution.
- **Alert logic:** Trigger email alerts only when `class_name in Config.ALERT_CLASSES` (i.e., `fire` or `smoke`).
- **Feedback buttons:** Update the dynamic feedback reporting to reflect the predicted class (e.g., "Report Incorrect: Not a Fire").

---

#### [MODIFY] [utils/alerts.py](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/utils/alerts.py)
- Update `send_email_alert()` to accept and display the predicted class name and confidence in both plain-text and HTML email bodies.
- Update the subject line to include the class: e.g. `🔥 Fire Detected — SmokeSignal‑AI` or `💨 Smoke Detected — SmokeSignal‑AI`.
- Fix the syntax typo on line 194 (`marginThank you for your time...` → `margin:0;`).

---

#### [MODIFY] [utils/preprocess.py](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/utils/preprocess.py)
- No functional changes needed — the preprocessing pipeline (resize + normalize to [0,1]) is architecture-agnostic and works for both binary and multi-class models.

---

#### [MODIFY] [tests/test_ood.py](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/tests/test_ood.py)
- Update prediction extraction from single sigmoid to multi-class argmax:
  ```python
  # Before
  wildfire_conf = float(preds[0][0])

  # After
  class_idx = int(np.argmax(preds[0]))
  class_name = Config.CLASSES[class_idx]
  class_conf = float(preds[0][class_idx])
  ```
- Update the results table to show predicted class name and per-class confidence.

---

#### [MODIFY] [benchmark_inference.py](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/benchmark_inference.py)
- Update the model path to load the multi-class model.
- No changes to benchmarking logic (inference shape is the same).

---

#### [MODIFY] [GEMINI.md](file:///c:/Users/dhyan/Documents/DP%20Code's/Machine%20Learning/Projects/SmokeSignal-AI/GEMINI.md)
- Append changelog entry for issue #12 work.

---

## Phase 4 — Verification & Acceptance

### Automated Tests
```bash
# 1. Curate and verify the edge-case dataset
python src/curate_dataset.py

# 2. Train the multi-class model (local)
python src/train_multiclass.py

# 3. Run OOD robustness tests against the new model
python tests/test_ood.py
```

### Acceptance Criteria Checklist

| # | Criterion | How Verified |
|---|---|---|
| 1 | ≥500 sunset + ≥500 dark cloud images in validation set | `curate_dataset.py` prints count summary |
| 2 | FPR < 5% on sunset and dark cloud edge-case sets | Regression check in `train_multiclass.py` |
| 3 | Model outputs 5-class predictions with confidence | Verify model output shape is `(None, 5)` |
| 4 | Precision/recall on existing benchmark ≤ 2% regression | Automated comparison in `train_multiclass.py` |
| 5 | Inference pipeline propagates class labels + confidence | Manual test via `streamlit run app.py` |
| 6 | Model weights + class labels committed | `model/wildfire_detector_multiclass.keras` + `model/class_labels.json` |

### Manual Verification
1. Launch `streamlit run app.py` and upload:
   - A sunset/sunrise image → should predict `clear`, **no** email alert.
   - A dark monsoon cloud image → should predict `cloud`, **no** email alert.
   - A wildfire satellite image → should predict `fire`, email alert **sent**.
   - A smoky scene → should predict `smoke`, email alert **sent**.
2. Verify email content includes the predicted class name and confidence score.
